cmd/dagger/check.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	var printCheck func(*dagger.Check) error
+	printCheck = func(check *dagger.Check) error {
+		name, err := check.Name(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get check name: %w", err)
+		}
+		name = strcase.ToKebab(name)
+
+		descr, err := check.Description(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get check description: %w", err)
+		}
+		fmt.Fprintf(tw, "%s\t%s\n", name, descr)
+		return nil
+	}
+
+	for _, check := range envChecks {
+		// TODO: this shouldn't be needed, there is a bug in our codegen for lists of objects. It should
+		// internally be doing this so it's not needed explicitly
+		checkID, err := check.ID(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get check id: %w", err)
+		}
+
+		check = *c.Check(dagger.CheckOpts{ID: checkID})
+		err = printCheck(&check)
+		if err != nil {
+			return err
+		}
+	}
+
+	return tw.Flush()
+}
+
+func RunCheck(ctx context.Context, _ *client.Client, c *dagger.Client, env *dagger.Environment, cmd *cobra.Command, dynamicCmdArgs []string) (err error) {
+	rec := progrock.RecorderFromContext(ctx)
+
+	// TODO(vito): this is pretty confusing, but we need to initialize a root
+	// group for subsequent groups + vertices
+	rec = rec.WithGroup(progrock.RootGroup)
+	ctx = progrock.RecorderToContext(ctx, rec)
+
+	var selectedChecks []*dagger.Check
+	if len(dynamicCmdArgs) > 0 {
+		for _, checkName := range dynamicCmdArgs {
+			// we accept check names in both kebab-case and lowerCamelCase
+			checkName = strcase.ToLowerCamel(checkName)
+			selectedChecks = append(selectedChecks, env.Check(checkName))
+		}
+	} else {
+		envChecks, err := env.Checks(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get environment commands: %w", err)
+		}
+		// TODO: workaround bug in codegen
+		for _, check := range envChecks {
+			check := check
+			checkID, err := check.ID(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to get check id: %w", err)
+			}
+			selectedChecks = append(selectedChecks, c.Check(dagger.CheckOpts{ID: checkID}))
+		}
+	}
+
+	path := []string{}
+	var eg errgroup.Group
+	for _, check := range selectedChecks {
+		check := check
+		runCheckHierarchy(ctx, c, rec, path, &eg, check)
+	}
+	return eg.Wait()
+}
+
+func runCheckHierarchy(
+	ctx context.Context,
+	c *dagger.Client,
+	rec *progrock.Recorder,
+	path []string,
+	eg *errgroup.Group,
+	check *dagger.Check,
+) error {
+	name, err := check.Name(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get check name: %w", err)
+	}
+	description, err := check.Description(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get check description: %w", err)
+	}
+
+	if name == "" {
+		name = description
+	}
+	if name == "" {
+		name = identity.NewID()
+	}
+	name = strcase.ToKebab(name)
+
+	path = append([]string{}, path...)
+	path = append(path, name)
+	fullPathName := strings.Join(path, "->")
+	digest := digest.FromString(fullPathName)
```

Comment:
It's tempting to use the check ID here. Or can the same check ID be present in multiple places in one `checks` call?

Reply:
Yeah, and using the full path name is probably not needed after your fixes; that whole approach was just part of my original naive attempt at getting the progress output hierarchy working.

`CheckID` includes the environment ID (to enable invoking the check), so I think it should be safe to use... Will double check and update if so